[wip] prototyping for rangefeed scalability work

wenyihu6 opened the pull request:
Just prototyping - This patch reduces registration goroutines and removes r.buf at rangefeed level.

TODO: 
- rationalize memory accounting
- tests 
- is it okay to send same registration error twice 

Shutdown logic: https://docs.google.com/document/d/1rQzVYnI4XuWloenEYa_9cvQ5aVYPlhR2cb8IAQKo6_M/edit?usp=sharing

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124757)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver/rangefeed: refactor and add stream muxer

This patch refactors the current rangefeed completion watcher to use stream muxer. Note that the main purpose is to make future commits cleaner. We will get rid of future pkg in the next commit.

wenyihu6 committed:
add activestreams to stream muxer

wenyihu6 committed:
delegate metrics updates to streammuxer

wenyihu6 committed:
remove future pkg

a

wenyihu6 committed:
removed locked rangefeed stream

wenyihu6 committed:
remove non-muxrangefeed metrics

wenyihu6 committed:
a

wenyihu6 committed:
add rangefeed level cleanup good addmuxer code

wenyihu6 committed:
add event queue

wenyihu6 committed:
add node level shutdown clean up

wenyihu6 committed:
rename

wenyihu6 committed:
changed shutdown a bit

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
change to use rangeID from map

wenyihu6 committed:
a

wenyihu6 committed:
added non buffered registry

wenyihu6 commented:
Was just for prototyping.
